Fix file clean up problem

kursatkobya opened the pull request.

kursatkobya committed:
Fix file clean up problem

tobsan commented on common/filetoolkitwithundo.cpp:
```diff
@@ -226,19 +226,32 @@ ReturnCode FileToolkitWithUndo::createSharedMountPoint(const std::string &path)
     return ReturnCode::SUCCESS;
 }
 
+bool FileToolkitWithUndo::isPathExist(const std::string path)
```
The name of this function isn't very obvious. hasPathAlready or something would make more sense.

tobsan commented on common/filetoolkitwithundo.cpp:
```diff
@@ -226,19 +226,32 @@ ReturnCode FileToolkitWithUndo::createSharedMountPoint(const std::string &path)
     }
-    m_cleanupHandlers.push(new FileCleanUpHandler(path));
+
+    if (!isPathExist(path)) {
```
Shouldn't this type of check be done for any cleanup handler? I mean it's good that it is being done here, but shouldn't be done for mounts etc also, to make it more generic?

Perhaps this is a bit over-the-top though, since there is no common way to compare cleanup handlers :/

kursatkobya replied:
make sense, will do

tobsan commented on common/filetoolkitwithundo.h:
```diff
@@ -112,12 +112,19 @@ class FileToolkitWithUndo
      *  FileToolKitWithUndo that will be run from the destructor.
      */
-    std::stack<CleanUpHandler *> m_cleanupHandlers;
+    std::vector<CleanUpHandler *> m_cleanupHandlers;
```
Why is this changed from a stack to a vector? The desired behavior is LIFO as a stack, so why not stick with it?

kursatkobya replied:
searching en element in stack does not meaningful since it does not support random access

kursatkobya committed:
Create generic function to compare cleanup handlers

Add a pure virtual function called queryName to the CleanUpHandler class

Fill queryName function in all derived classes

Add querry before each addition to cleanuphandler list

tobsan commented on common/overlaysynccleanuphandler.cpp:
```diff
@@ -32,3 +32,8 @@ ReturnCode OverlaySyncCleanupHandler::clean()
+
+const std::string OverlaySyncCleanupHandler::queryName()
+{
+    return "";
```
This does not make sense. I would say that the destination path would be what we want to check here? The reason for that would be that it is valid to mount the same location to many places (in theory), but not to mount several locations to the same place.

tobsan reviewed (changes requested):
Just one comment on the OverlaySyncCleanupHandler

tobsan commented on common/overlaysynccleanuphandler.h:
```diff
@@ -43,6 +43,15 @@ class OverlaySyncCleanupHandler : public CleanUpHandler
+    /**
+     * @brief this function is needed to query member name yet it is not needed for
+     * OverlaySyncCleanupHandler. Since OverlaySyncCleanupHandler uses DirectoryCleanUpHandler
+     * for removing directory, this function won't be called ever.
```
1. Only use one short sentence for @brief, write the longer description in a separate paragraph.
2. This function **is** called when we iterate over the cleanuphandlers, but the reasoning is good anyway.

tobsan commented on common/overlaysynccleanuphandler.h:
```diff
@@ -43,6 +43,17 @@ class OverlaySyncCleanupHandler : public CleanUpHandler
+     *
+     * @return an empty string
+     */
+    const std::string queryName() override;
```
Sorry, but I still don't like this description. OverlaySyncCleanupHandler doesn't use DirectoryCleanupHandler, it uses RecursiveCopy.

The reason why we're returning an empty string here is because we don't want any paths to compare equal to an overlay path, isn't it? Then we should write that.

kursatkobya replied:
Actually we are returning empty string because no directory removal operation occurs within OverlaySyncCleanupHandler, so it is irrelevant when the code checks whether the directory is in the list or not since we are doing this for cleaning path. Maybe i should add this as a comment.

tobsan replied:
Sure, but this check is done for anything we want to add a cleanup handler for, so it is a little more generic than just directories.

DunderRoffe reviewed (changes requested):
Sorry to be picky but there should really be a description attached to the commit message. This is in order to make it easier when searching through the commits at a later stage.

kursatkobya committed:
Adjust function localizations in FileToolkitWithUndo class

Adjust process order of CleanupHandlers
